### PR TITLE
blog post on interaction patterns

### PR DESCRIPTION
**Pull request by autodidaddict**
New blog post. No image yet.

### CODE REVIEW

File: blog/webassembly_interaction_patterns.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+draft: false
+---
+The first thing people need to learn when encountering WebAssembly for the first time is how to interact with the code inside the module.
+There are a number of patterns available for this, and in this blog post I'll cover the main three: _Command, Reactor_, and _Library_.
+
+<!-- truncate -->
+
+As we've mentioned in a number of posts and conference talks, WebAssembly's isolated sandbox forces people to use functions that
+can only accept and return numbers. _Bailey Hayes_ likes to refer to this aspect of WebAssembly as just _"[a bunch of numbers in a trenchcoat](https://youtu.be/6_BRLqxiZPU?t=459)"_. There are 
+plenty of libraries, wrappers, tools, and extensions that try and hide this basic fact, but it's important to understand underneath all of that, it's comes down to exchanging numbers.
```
Comment (aschrijver) [inline blog/webassembly_interaction_patterns.md]: Typo: "it comes down to"

File: blog/webassembly_interaction_patterns.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+## Summary
+In this blog post, we took a quick tour of the three main patterns for working with WebAssembly modules. We saw that the command pattern is the simplest and most portable, but it's also the most limited. The reactor pattern is a bit more complex, but it's also more flexible. The library pattern is the most complex (this should change in the near future), but it's also the most flexible. If we distill things down to their smallest cores, we can equate these WebAssembly interaction patterns to classic strategies most of us are already familiar with:
+
+* _**Command** -> Serverless_
+* _**Reactor** -> Services_
+* _**Library** -> Components & Hybrid_
```
Comment (brooksmtownsend) [inline blog/webassembly_interaction_patterns.md]: I think it would be cool to wrap it up here by nailing what wasmCloud's interaction pattern is! 😄

**Review by brooksmtownsend — COMMENTED** [review]
This is a cool post, left one comment but it's good to go with an image I think

File: blog/webassembly_interaction_patterns.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+--image: "/img/there-and-back-again.jpg"
+date: 2023-06-06T09:00:00+05:00
+author: "Kevin Hoffman"
+author_profile: "https://mastodon.world/@autodidaddict"
+categories: ["webassembly", "components", "patterns"]
+description: Exploring WebAssembly Interaction Patterns - Command, Reactor, and Library"
+draft: false
+---
+The first thing people need to learn when encountering WebAssembly for the first time is how to interact with the code inside the module.
+There are a number of patterns available for this, and in this blog post I'll cover the main three: _Command, Reactor_, and _Library_.
```
Comment (CazJT) [inline blog/webassembly_interaction_patterns.md]: Adjust comma and add comma:

available for this and, in this blog post,

File: blog/webassembly_interaction_patterns.md (hunk 1 above)
Comment (CazJT) [inline blog/webassembly_interaction_patterns.md]: Remove the Oxford comm after tools

Comment (CazJT) [inline blog/webassembly_interaction_patterns.md]: Add comma before underneath:

There are plenty of libraries, wrappers, tools and extensions that try and hide this basic fact, but it's important to understand, underneath it all, it comes down to exchanging numbers.

File: blog/webassembly_interaction_patterns.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+Once we get past the fact that we're only dealing with numbers, we can then move on to explore _how_ we're going to interact with the module. Let's take a look at each of these patterns in turn.
+
+## The Command Pattern
+In this pattern, a `command` is a bundle of logically connected information. This command represents an _imperative_: a request that work be performed. Generally
+speaking, the command pattern involves sending a request and then obtaining a response. The defining characteristics of this pattern are the single execution of the command and
+the request/response nature of the interaction.
+
+Commands can be anything from a simple request to add two numbers together to a complex request to perform object recognition on an image frame. The mechanics of getting this 
+command into a WebAssembly module and getting the response back usually involves building the module to target **WASI** and accepting a single command input over **stdin** and delivering
```
Comment (CazJT) [inline blog/webassembly_interaction_patterns.md]: Comma after WASI and consider losing the first 'and':

The mechanics of getting this command into a WebAssembly module and getting the response back usually involves building the module to target WASI, accepting a single command input over stdin and delivering the response via stdout.

File: blog/webassembly_interaction_patterns.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+the response via **stdout**.
+
+The primary appeal of this pattern is simplicity and availability. Even languages that struggle to produce freestanding WebAssembly modules (e.g. targeting `wasm32-unknown-unknown`) can often 
+produce `wasm32-wasi` modules. Accepting data via **stdin** means that the WASI plumbing is taking care of marshaling the data through linear memory and the developer doesn't have to do that manually. Returning the result is a simple matter of writing to **stdout**.
```
Comment (CazJT) [inline blog/webassembly_interaction_patterns.md]: Adjust to 'takes' care of marshaling; change and to so; change that to it:

Accepting data via stdin means that the WASI plumbing takes care of marshaling the data through the linear memory so the developer doesn't have to do it manually.

File: blog/webassembly_interaction_patterns.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+The net effect on the developer is a huge boon during today's highly volatile Wasm ecosystem. You can write code that runs both as a regular console app and as a WASI module. This makes
+the code easy to reason about, easy to build, easy to test, and (theoretically) automatically portable to any platform that supports WASI. Sending the request to the module is a simple matter of invoking its default exported function (usually `start` or `_start`).
```
Comment (CazJT) [inline blog/webassembly_interaction_patterns.md]: Remove Oxford comma after east to test

File: blog/webassembly_interaction_patterns.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+## The Reactor Pattern
+It's easy to overthink this one. Rather than being named after some specific library or technology, in this case **reactor** is as its name implies: it _reacts_. 
+The reactor pattern is a command pattern that doesn't shut down after processing a single request. Instead, it stays running and waits for more requests. While both the command and reactor patterns can make use of callbacks for host facilities, these types of callbacks are far more prevalent in reactors. Many implementations of the reactor pattern require that some form of `setup` or `init` function be called prior to the reactor being able to process requests.
+
+Another fundamental difference is that while the command pattern commonly maps a single module to a single function, a reactor can respond to multiple different types of stimulus (which can be modeled as commands, events, or anything else). You'll typically see a dispatcher inside a reactor that routes the request to the appropriate handler. This is a common pattern in many languages and frameworks, but it's especially common in JavaScript and Node.js.
```
Comment (CazJT) [inline blog/webassembly_interaction_patterns.md]: Comma between that and while - difference is that, while the command pattern....

**Review by CazJT — DISMISSED** [review]

**Review by brooksmtownsend — APPROVED** [review]

**Review by brooksmtownsend — APPROVED** [review]

**Pull request merged**